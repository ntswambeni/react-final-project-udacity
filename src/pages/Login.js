import { connect } from "react-redux";
import { handleSetAuthedUser } from "../actions/authedUser";
import { useState } from "react";
import { Navigate, useLocation } from "react-router-dom";

const Login = (props) => {
  const [formfields, setFormfields] = useState({ username: "", password: "" });
  const { username, password } = formfields;

  const location = useLocation();

  const handleChange = (e) => {
    setFormfields((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));
  };

  const handleLogin = (e) => {
    e.preventDefault();
    props.dispatch(handleSetAuthedUser(formfields));
    setFormfields({ username: "", password: "" });
  };

  if (props.authedUser) {
    if (location.state) {
      return <Navigate to={location.state.from} replace />;
    }
    return <Navigate to="/" replace />;
  }

  return (
    <div className="login-page container">
      <h1>Employee Polls</h1>
      <img
        className="login-page__cover"
        src="https://www.applauz.me/hubfs/image%20%2822%29.png"
        alt="Login page cover"
      ></img>
      <h2>Log In</h2>
      <form className="form">
        <label className="form__label">
          User
          <input
            type="text"
            name="username"
            className="form__input"
            placeholder="User"
            data-testid="username"
            onChange={handleChange}
          />
        </label>
        <label className="form__label">
          Password
          <input
            type="password"
            name="password"
            className="form__input"
            placeholder="Password"
            data-testid="password"
            onChange={handleChange}
          />
        </label>
        <button
          className="form__button"
          type="submit"
          onClick={handleLogin}
          disabled={username === "" || password === ""}
        >
          Login
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authedUser,
});

export default connect(mapStateToProps)(Login);
