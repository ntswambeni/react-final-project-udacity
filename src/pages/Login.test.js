import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { configureStore } from "@reduxjs/toolkit";
import rootReducers from "../reducers";
import middleware from "../middleware";
import Login from "./Login";

const store = configureStore({ reducer: rootReducers, middleware: middleware });
describe("Login", () => {
  it("matchs snapshot", () => {
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    expect(component).toMatchSnapshot();
  });

  it("displays username and password fields and a button", () => {
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    const username = component.getByTestId("username");
    const password = component.getByTestId("password");
    const submit = component.getByTestId("submit");

    expect(username).toBeInTheDocument();
    expect(password).toBeInTheDocument();
    expect(submit).toBeInTheDocument();
  });

  it("clears the form after submit", () => {
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    const username = component.getByTestId("username");
    const password = component.getByTestId("password");
    const submit = component.getByTestId("submit");
    const alertMock = jest.spyOn(window, "alert").mockImplementation();

    fireEvent.change(username, { target: { value: "sarahado" } });
    fireEvent.change(password, { target: { value: "password" } });
    fireEvent.click(submit);
    expect(username.value).toEqual("");
    expect(password.value).toEqual("");
  });
});
