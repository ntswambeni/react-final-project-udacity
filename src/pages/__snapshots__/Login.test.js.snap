// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matchs snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-page container"
      >
        <h1>
          Employee Polls
        </h1>
        <img
          alt="Login page cover"
          class="login-page__cover"
          src="https://www.applauz.me/hubfs/image%20%2822%29.png"
        />
        <h2>
          Log In
        </h2>
        <form
          class="form"
        >
          <label
            class="form__label"
          >
            User
            <input
              class="form__input"
              data-testid="username"
              name="username"
              placeholder="User"
              type="text"
              value=""
            />
          </label>
          <label
            class="form__label"
          >
            Password
            <input
              class="form__input"
              data-testid="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </label>
          <button
            class="form__button"
            data-testid="submit"
            disabled=""
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-page container"
    >
      <h1>
        Employee Polls
      </h1>
      <img
        alt="Login page cover"
        class="login-page__cover"
        src="https://www.applauz.me/hubfs/image%20%2822%29.png"
      />
      <h2>
        Log In
      </h2>
      <form
        class="form"
      >
        <label
          class="form__label"
        >
          User
          <input
            class="form__input"
            data-testid="username"
            name="username"
            placeholder="User"
            type="text"
            value=""
          />
        </label>
        <label
          class="form__label"
        >
          Password
          <input
            class="form__input"
            data-testid="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </label>
        <button
          class="form__button"
          data-testid="submit"
          disabled=""
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
